cloudsmith/resource_webhook.go

```diff
@@ -0,0 +1,454 @@
+package cloudsmith
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cloudsmith-io/cloudsmith-api-go"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/samber/lo"
+)
+
+var (
+	eventTypes = []string{
+		"*",
+		"package.created",
+		"package.deleted",
+		"package.downloaded",
+		"package.failed",
+		"package.security_scanned",
+		"package.synced",
+		"package.syncing",
+		"package.tags_updated",
+	}
+	requestBodyFormatMap = map[int]string{
+		0: "JSON Object",
+		1: "JSON Array",
+		2: "Form Encoded JSON Object",
+		3: "Handlebars Template",
+	}
+	requestBodyTemplateFormatMap = map[int]string{
+		0: "Generic (user-defined)",
+		1: "JSON (application/json)",
+		2: "XML (application/xml)",
+	}
+)
+
+// expandEvents extracts "events" from TF state as a *schema.Set and converts to
+// a slice of strings we can use when interacting with the Cloudsmith API.
+func expandEvents(d *schema.ResourceData) []string {
+	set := d.Get("events").(*schema.Set)
+	return lo.Map(set.List(), func(item interface{}, _ int) string {
+		return item.(string)
+	})
+}
+
+// flattenEvents takes a slice of strings as returned by the Cloudsmith API and
+// converts to a *schema.Set that can be stored in TF state.
+func flattenEvents(events []string) *schema.Set {
+	set := schema.NewSet(schema.HashString, []interface{}{})
+	for _, event := range events {
+		set.Add(event)
+	}
+	return set
+}
+
+// expandRequestBodyFormat extracts the request body format from TF state as a
+// human-readable string (if set) and converts it to an int64 that can be used
+// to interact with the Cloudsmith API.
+func expandRequestBodyFormat(d *schema.ResourceData) *int64 {
+	value := optionalString(d, "request_body_format")
+	if value == nil {
+		return nil
+	}
+
+	m := lo.Invert(requestBodyFormatMap)
+	v, ok := m[*value]
+
+	if !ok {
+		return nil
+	}
+
+	return cloudsmith.PtrInt64(int64(v))
+}
+
+// flattenRequestBodyFormat takes an int64 as returned by the Cloudsmith API and
+// converts it to a human readable string that can be stored/used in TF state.
+func flattenRequestBodyFormat(fmt int64) string {
+	return requestBodyFormatMap[int(fmt)]
+}
+
+// expandRequestBodyTemplateFormat extracts the request body template format
+// from TF state as a human-readable string (if set) and converts it to an int64
+// that can be used to interact with the Cloudsmith API.
+func expandRequestBodyTemplateFormat(d *schema.ResourceData) *int64 {
```

Comment:
this looks similar to `expandRequestBodyFormat` above, is there a reason we couldn't combine the two and pass in the formatMap as a parameter?

Reply:
no major reason we couldn't, but i thought leaving it explicit for now was better as i think it may need changed in the very bear future